src/test/crimson/test_perfcounters.cc

```diff
@@ -0,0 +1,59 @@
+#include <pthread.h>
+#include <iostream>
+#include <fmt/format.h>
+#include "common/Formatter.h"
+#include <seastar/core/app-template.hh>
+#include <seastar/core/sharded.hh>
+#include "common/perf_counters.h"
+#include "crimson/common/perf_counters_service.h"
+
+using ceph::common::PerfCountersService;
+enum {
+  perftest_first = 1000000,
+  perftest_count,
+  perftest_last,
+};
+
+static seastar::future<> test_perfcounters(){
+  return ceph::common::sharded_pfservice().start().then([] {
+    return ceph::common::sharded_pfservice().invoke_on_all([] (auto& s){
+      PerfCounters *perf_logger;
+      //char name[128];
+      //sprintf(name, "seastar-osd::shard-%ud", engine().cpu_id());
+      std::string name = fmt::format("seastar-osd::shard-{id}",fmt::arg("id",seastar::engine().cpu_id()));
+      PerfCountersBuilder plb(NULL, name, perftest_first, perftest_last);
+      plb.add_u64_counter(perftest_count, "perftest_count", "count perftest");
+      perf_logger = plb.create_perf_counters();
+      perf_logger->inc(perftest_count);
+      s.perf_collection->add(perf_logger);
+    });
+  }).then([] {
+    return ceph::common::sharded_pfservice().invoke_on_all([] (auto& s){
+      ostringstream ostr;
+      Formatter* f = Formatter::create("json-pretty", "json-pretty", "json-pretty");
+      s.perf_collection->dump_formatted(f, false);
+      ostr << "dump perfcounters";
```

Comment:
https://github.com/ceph/ceph/pull/24043#discussion_r217712529

Reply:
@tchaikov hi kefu, I don't understand what you mean here., could you please explain in detail?

Reply:
> might want to verify the "perftest_count" and fail the test it's not expected?

@liu-chunmei unless you are going to parse the `ostr.str()` using a JSON parse and check the perftest_count, i think a better way to test is to use `with_counters()`.

Reply:
> @tchaikov formatter is already set to json, and it can dump out the result.

@liu-chunmei i meant, we cannot rely on visual inspection for unit test.

Reply:
@liu-chunmei i'd suggest something like following patch, in which:

- capitalize the enum 
- define the incremental value as a random value instead of using the default "1", so it can be verified later.
- define the perf counter's name, so we can 
- simplify the `fmt::format()` call
- remove the continuation to print out the counters
- use `finally()` instead of `then()` for stopping `sharded_pfcollection()`

```diff
diff --git a/src/test/crimson/test_perfcounters.cc b/src/test/crimson/test_perfcounters.cc
index c71f01a114..cab3ffb684 100644
--- a/src/test/crimson/test_perfcounters.cc
+++ b/src/test/crimson/test_perfcounters.cc
@@ -8,45 +8,35 @@
 #include "crimson/common/perf_counters_collection.h"
 
 enum {
-  perftest_first = 1000000,
-  perftest_count,
-  perftest_last,
+  PERFTEST_FIRST = 1000000,
+  PERFTEST_INDEX,
+  PERFTEST_LAST,
 };
 
+static constexpr uint64_t PERF_VAL = 42;
+
 static seastar::future<> test_perfcounters(){
   return ceph::common::sharded_pfcollection().start().then([] {
-    return ceph::common::sharded_pfcollection().invoke_on_all([] (auto& s){
-      std::string name = fmt::format("seastar-osd::shard-{id}",fmt::arg("id",seastar::engine().cpu_id()));
-      PerfCountersBuilder plb(NULL, name, perftest_first, perftest_last);
-      plb.add_u64_counter(perftest_count, "perftest_count", "count perftest");
+    return ceph::common::sharded_pfcollection().invoke_on_all([] (auto& s) {
+      std::string name = fmt::format("seastar-osd::shard-{}", seastar::engine().cpu_id());
+      PerfCountersBuilder plb(NULL, name, PERFTEST_FIRST, PERFTEST_LAST);
+      plb.add_u64_counter(PERFTEST_INDEX, "perftest_count", "count perftest");
       auto perf_logger = plb.create_perf_counters();
-      perf_logger->inc(perftest_count);
+      perf_logger->inc(PERFTEST_INDEX, PERF_VAL);
       s.get_perf_collection()->add(perf_logger);
     });
-  }).then([] {
-    return ceph::common::sharded_pfcollection().invoke_on_all([] (auto& s){
-      ostringstream ostr;
-      Formatter* f = Formatter::create("json-pretty", "json-pretty", "json-pretty");
-      s.get_perf_collection()->dump_formatted(f, false);
-      ostr << "dump perfcounters";
-      f->flush(ostr);
-      std::cout <<  ostr.str() << std::endl;
-    });
   }).then([]{
     return ceph::common::sharded_pfcollection().invoke_on_all([] (auto& s){ 
       auto pcc = s.get_perf_collection();
-      pcc->with_counters([=](const PerfCountersCollectionImpl::CounterMap &by_path){
-        for (const auto &i : by_path) {
-          auto& path = i.first;
-          auto& data = *(i.second.data);
-          auto& perf_counters = *(i.second.perf_counters);
-          std::stringstream msg;
-          msg << "path is " << path << " value is "<<data.u64 << "\n"; 
-          std::cout << msg.str();
+      pcc->with_counters([=](auto& by_path){
+        for (const auto &[path, perf_counter_ref] : by_path) {
+          if (PERF_VAL != perf_counter_ref.perf_counters->get(PERFTEST_INDEX)) {
+            throw std::runtime_error("perf counter does not match");
+          }
          }
       });
     });
-  }).then([] {
+  }).finally([] {
      return ceph::common::sharded_pfcollection().stop();
   });
 
 ```